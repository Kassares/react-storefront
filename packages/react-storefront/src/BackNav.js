/**
 * @license
 * Copyright Â© 2017-2018 Moov Corporation.  All rights reserved.
 */
import React, { Component } from 'react'
import get from 'lodash/get'
import ArrowLeft from '@material-ui/icons/KeyboardArrowLeft'
import ListViewIcon from '@material-ui/icons/ViewAgenda'
import GridViewIcon from '@material-ui/icons/BorderAll'
import withStyles from '@material-ui/core/styles/withStyles'
import Typography from '@material-ui/core/Typography'
import { inject, observer } from 'mobx-react'
import PropTypes from 'prop-types'
import classnames from 'classnames'
import { LAYOUT_LIST, LAYOUT_GRID } from './model/SearchResultsModelBase'
import Paper from '@material-ui/core/Paper'

/**
 * A header that allows the user to navigate back.  When a searchResults prop is provided,
 * this component also allows the user to switch between grid and list views.
 */
export const styles = theme => ({
  root: {
    padding: '6px 86px',
    position: 'relative',
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
    borderTop: `1px solid ${theme.palette.divider}`,
    zIndex: theme.zIndex.appBar - 1,

    '& a': {
      color: theme.palette.text.primary,
      textDecoration: 'none'
    }
  },
  label: {
    fontWeight: 'bold',
    fontSize: '18px',
    display: 'block',
    textAlign: 'center',
    textTransform: 'uppercase',
    padding: 0
  },
  backButtonWrapper: {
    position: 'absolute',
    top: 0,
    left: '5px',
    height: '41px',
    display: 'flex',
    alignItems: 'center'
  },
  switchButtonsWrapper: {
    position: 'absolute',
    display: 'flex',
    alignItems: 'center',
    top: 0,
    right: '10px',
    height: '41px'
  },
  switchButton: {
    marginLeft: '5px',
    alignItems: 'center',
    display: 'flex',
    padding: '3px',
    borderRadius: '50%'
  },
  selectedSwitchButton: {
    backgroundColor: '#F4F2F1'
  }
})

@withStyles(styles, { name: 'RSFBackNav' })
@inject('history')
@observer
export default class BackNav extends Component {
  static propTypes = {
    /**
     * The text to display representing the previous location.
     */
    text: PropTypes.string.isRequired,

    /**
     * The url to navigate to when clicked.  If omitted, this component will navigate back in the history when clicked.
     */
    url: PropTypes.string,

    /**
     * When displaying this component on a search results page (such as a subcategory), you can supply the SearchResultsModelBase instance here
     * and this component will allow you to switch between grid and list views.
     */
    searchResults: PropTypes.shape({
      layout: PropTypes.string.isRequired
    }),

    /**
     * Overrides the default HTML element used for the label (span).
     */
    labelComponent: PropTypes.string
  }

  static defaultProps = {
    labelComponent: 'span'
  }

  switchLayout = layout => {
    const { searchResults } = this.props
    searchResults.switchLayout(layout)
  }

  renderViewToggle() {
    const { classes, searchResults } = this.props
    const layout = get(searchResults, 'layout', LAYOUT_GRID)

    return searchResults ? (
      <span className={classes.switchButtonsWrapper}>
        <span
          className={classnames(classes.switchButton, {
            [classes.selectedSwitchButton]: layout === LAYOUT_GRID
          })}
        >
          <GridViewIcon
            color={layout === LAYOUT_GRID ? 'secondary' : 'primary'}
            onClick={() => this.switchLayout(LAYOUT_GRID)}
          />
        </span>
        <span
          className={classnames(classes.switchButton, {
            [classes.selectedSwitchButton]: layout === LAYOUT_LIST
          })}
        >
          <ListViewIcon
            color={layout === LAYOUT_LIST ? 'secondary' : 'primary'}
            onClick={() => this.switchLayout(LAYOUT_LIST)}
          />
        </span>
      </span>
    ) : null
  }

  render() {
    const { text, classes, labelComponent } = this.props

    return (
      <Paper className={classes.root}>
        <Typography variant="caption">
          <span
            onClick={() => {
              this.onBack()
            }}
            className={classes.backButtonWrapper}
          >
            <ArrowLeft />
          </span>
          {React.createElement(labelComponent, {className: classes.label}, text)}
          {this.renderViewToggle()}
        </Typography>
      </Paper>
    )
  }

  onBack = () => {
    const { history, url } = this.props

    if (url) {
      history.push(url)
    } else {
      history.goBack()
    }
  }
}
