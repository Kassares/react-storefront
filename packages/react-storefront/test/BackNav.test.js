/**
 * @license
 * Copyright Â© 2017-2018 Moov Corporation.  All rights reserved.
 */
import React from 'react'
import { mount } from 'enzyme'
import { Provider } from 'mobx-react'
import { createMemoryHistory } from 'history'
import BackNav from '../src/BackNav'
import AppModelBase from '../src/model/AppModelBase'
import SearchResultsModelBase, { LAYOUT_LIST, LAYOUT_GRID } from '../src/model/SearchResultsModelBase'

describe('BackNav', () => {

  let app, history

  beforeEach(() => {
    app = AppModelBase.create({})
    history = createMemoryHistory({
      initialEntries: [
        '/c/2'
      ]
    })
  })

  it('renders', () => {
    expect(mount(
      <Provider history={history} app={app}>
        <BackNav text="Rugs" url="/s/1" />
      </Provider>
    )).toMatchSnapshot()
  })

  it('renders the view type switcher', () => {
    const searchResults = SearchResultsModelBase.create({ id: "1" })

    expect(mount(
      <Provider history={history}>
        <BackNav text="Furniture" url="/c/1" searchResults={searchResults} />
      </Provider>
    )).toMatchSnapshot()
  })


  it('render the corresponding tag if `labelComponent` prop provided', () => {
    expect(mount(
      <Provider history={history}>
        <BackNav text="Rugs" url="/rugs/1" labelComponent="h1" />
      </Provider>
    )).toMatchSnapshot()
  })

  it('goes to the specified url when clicked', () => {
    const wrapper = mount(
      <Provider history={history}>
        <BackNav text="Rugs" url="/c/1" />
      </Provider>
    )

    wrapper.find('span[onClick]').at(0).simulate('click')
    expect(history.location.pathname).toBe('/c/1')
  })

  it('goes back in history when no url is provided', () => {
    const wrapper = mount(
      <Provider history={history}>
        <BackNav text="Back"/>
      </Provider>
    )

    wrapper.find('span[onClick]').at(0).simulate('click')
    expect(history.location.pathname).toBe('/c/2')
  })

  it('switches the view types to grid when clicked', () => {
    const searchResults = SearchResultsModelBase.create({ id: "1", layout: LAYOUT_LIST })

    const wrapper = mount(
      <Provider history={history}>
        <BackNav text="Back" searchResults={searchResults}/>
      </Provider>
    )

    wrapper.find('BorderAllIcon[onClick]').at(0).simulate('click')
    expect(searchResults.layout).toBe(LAYOUT_GRID)
  })

  it('switches the view types to list when clicked', () => {
    const searchResults = SearchResultsModelBase.create({ id: "1", layout: LAYOUT_GRID })

    const wrapper = mount(
      <Provider history={history}>
        <BackNav text="Back" searchResults={searchResults}/>
      </Provider>
    )

    wrapper.find('ViewAgendaIcon[onClick] svg').at(0).simulate('click')
    expect(searchResults.layout).toBe(LAYOUT_LIST)
  })
})