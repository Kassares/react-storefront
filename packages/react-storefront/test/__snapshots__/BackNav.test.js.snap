// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BackNav renders 1`] = `
<Provider history={{...}} app={{...}}>
  <WithStyles(inject-BackNav-with-history) text="Rugs" url="/s/1">
    <inject-BackNav-with-history text="Rugs" url="/s/1" classes={{...}}>
      <BackNav text="Rugs" url="/s/1" classes={{...}} history={{...}}>
        <WithStyles(Paper) className="RSFBackNav-root-1">
          <Paper className="RSFBackNav-root-1" classes={{...}} component="div" elevation={2} square={false}>
            <div className="MuiPaper-root-7 MuiPaper-elevation2-11 MuiPaper-rounded-8 RSFBackNav-root-1">
              <WithStyles(Typography) variant="caption">
                <Typography variant="caption" theme={{...}} classes={{...}} align="inherit" color="default" gutterBottom={false} headlineMapping={{...}} noWrap={false} paragraph={false}>
                  <span className="MuiTypography-root-34 MuiTypography-caption-44">
                    <span onClick={[Function: onClick]} className="RSFBackNav-backButtonWrapper-3">
                      <pure(KeyboardArrowLeftIcon)>
                        <KeyboardArrowLeftIcon>
                          <WithStyles(SvgIcon)>
                            <SvgIcon classes={{...}} color="inherit" component="svg" fontSize="default" viewBox="0 0 24 24">
                              <svg className="MuiSvgIcon-root-69" focusable="false" viewBox="0 0 24 24" color={[undefined]} aria-hidden="true" role="presentation">
                                <path d="M15.41 16.59L10.83 12l4.58-4.59L14 6l-6 6 6 6 1.41-1.41z" />
                                <path fill="none" d="M0 0h24v24H0V0z" />
                              </svg>
                            </SvgIcon>
                          </WithStyles(SvgIcon)>
                        </KeyboardArrowLeftIcon>
                      </pure(KeyboardArrowLeftIcon)>
                    </span>
                    <span className="RSFBackNav-label-2">
                      Rugs
                    </span>
                  </span>
                </Typography>
              </WithStyles(Typography)>
            </div>
          </Paper>
        </WithStyles(Paper)>
      </BackNav>
    </inject-BackNav-with-history>
  </WithStyles(inject-BackNav-with-history)>
</Provider>
`;

exports[`BackNav renders the view type switcher 1`] = `
<Provider history={{...}}>
  <WithStyles(inject-BackNav-with-history) text="Furniture" url="/c/1" searchResults={{...}}>
    <inject-BackNav-with-history text="Furniture" url="/c/1" searchResults={{...}} classes={{...}}>
      <BackNav text="Furniture" url="/c/1" searchResults={{...}} classes={{...}} history={{...}}>
        <WithStyles(Paper) className="RSFBackNav-root-1">
          <Paper className="RSFBackNav-root-1" classes={{...}} component="div" elevation={2} square={false}>
            <div className="MuiPaper-root-7 MuiPaper-elevation2-11 MuiPaper-rounded-8 RSFBackNav-root-1">
              <WithStyles(Typography) variant="caption">
                <Typography variant="caption" theme={{...}} classes={{...}} align="inherit" color="default" gutterBottom={false} headlineMapping={{...}} noWrap={false} paragraph={false}>
                  <span className="MuiTypography-root-34 MuiTypography-caption-44">
                    <span onClick={[Function: onClick]} className="RSFBackNav-backButtonWrapper-3">
                      <pure(KeyboardArrowLeftIcon)>
                        <KeyboardArrowLeftIcon>
                          <WithStyles(SvgIcon)>
                            <SvgIcon classes={{...}} color="inherit" component="svg" fontSize="default" viewBox="0 0 24 24">
                              <svg className="MuiSvgIcon-root-69" focusable="false" viewBox="0 0 24 24" color={[undefined]} aria-hidden="true" role="presentation">
                                <path d="M15.41 16.59L10.83 12l4.58-4.59L14 6l-6 6 6 6 1.41-1.41z" />
                                <path fill="none" d="M0 0h24v24H0V0z" />
                              </svg>
                            </SvgIcon>
                          </WithStyles(SvgIcon)>
                        </KeyboardArrowLeftIcon>
                      </pure(KeyboardArrowLeftIcon)>
                    </span>
                    <span className="RSFBackNav-label-2">
                      Furniture
                    </span>
                    <span className="RSFBackNav-switchButtonsWrapper-4">
                      <span className="RSFBackNav-switchButton-5 RSFBackNav-selectedSwitchButton-6">
                        <pure(BorderAllIcon) color="secondary" onClick={[Function: onClick]}>
                          <BorderAllIcon color="secondary" onClick={[Function: onClick]}>
                            <WithStyles(SvgIcon) color="secondary" onClick={[Function: onClick]}>
                              <SvgIcon color="secondary" onClick={[Function: onClick]} classes={{...}} component="svg" fontSize="default" viewBox="0 0 24 24">
                                <svg className="MuiSvgIcon-root-69 MuiSvgIcon-colorSecondary-71" focusable="false" viewBox="0 0 24 24" color={[undefined]} aria-hidden="true" role="presentation" onClick={[Function: onClick]}>
                                  <path d="M3 3v18h18V3H3zm8 16H5v-6h6v6zm0-8H5V5h6v6zm8 8h-6v-6h6v6zm0-8h-6V5h6v6z" />
                                  <path fill="none" d="M0 0h24v24H0z" />
                                </svg>
                              </SvgIcon>
                            </WithStyles(SvgIcon)>
                          </BorderAllIcon>
                        </pure(BorderAllIcon)>
                      </span>
                      <span className="RSFBackNav-switchButton-5">
                        <pure(ViewAgendaIcon) color="primary" onClick={[Function: onClick]}>
                          <ViewAgendaIcon color="primary" onClick={[Function: onClick]}>
                            <WithStyles(SvgIcon) color="primary" onClick={[Function: onClick]}>
                              <SvgIcon color="primary" onClick={[Function: onClick]} classes={{...}} component="svg" fontSize="default" viewBox="0 0 24 24">
                                <svg className="MuiSvgIcon-root-69 MuiSvgIcon-colorPrimary-70" focusable="false" viewBox="0 0 24 24" color={[undefined]} aria-hidden="true" role="presentation" onClick={[Function: onClick]}>
                                  <path fill="none" d="M0 0h24v24H0z" />
                                  <path d="M20 13H3c-.55 0-1 .45-1 1v6c0 .55.45 1 1 1h17c.55 0 1-.45 1-1v-6c0-.55-.45-1-1-1zm0-10H3c-.55 0-1 .45-1 1v6c0 .55.45 1 1 1h17c.55 0 1-.45 1-1V4c0-.55-.45-1-1-1z" />
                                </svg>
                              </SvgIcon>
                            </WithStyles(SvgIcon)>
                          </ViewAgendaIcon>
                        </pure(ViewAgendaIcon)>
                      </span>
                    </span>
                  </span>
                </Typography>
              </WithStyles(Typography)>
            </div>
          </Paper>
        </WithStyles(Paper)>
      </BackNav>
    </inject-BackNav-with-history>
  </WithStyles(inject-BackNav-with-history)>
</Provider>
`;

exports[`BackNav render the corresponding tag if \`labelComponent\` prop provided 1`] = `
<Provider history={{...}}>
  <WithStyles(inject-BackNav-with-history) text="Rugs" url="/rugs/1" labelComponent="h1">
    <inject-BackNav-with-history text="Rugs" url="/rugs/1" labelComponent="h1" classes={{...}}>
      <BackNav text="Rugs" url="/rugs/1" labelComponent="h1" classes={{...}} history={{...}}>
        <WithStyles(Paper) className="RSFBackNav-root-1">
          <Paper className="RSFBackNav-root-1" classes={{...}} component="div" elevation={2} square={false}>
            <div className="MuiPaper-root-7 MuiPaper-elevation2-11 MuiPaper-rounded-8 RSFBackNav-root-1">
              <WithStyles(Typography) variant="caption">
                <Typography variant="caption" theme={{...}} classes={{...}} align="inherit" color="default" gutterBottom={false} headlineMapping={{...}} noWrap={false} paragraph={false}>
                  <span className="MuiTypography-root-34 MuiTypography-caption-44">
                    <span onClick={[Function: onClick]} className="RSFBackNav-backButtonWrapper-3">
                      <pure(KeyboardArrowLeftIcon)>
                        <KeyboardArrowLeftIcon>
                          <WithStyles(SvgIcon)>
                            <SvgIcon classes={{...}} color="inherit" component="svg" fontSize="default" viewBox="0 0 24 24">
                              <svg className="MuiSvgIcon-root-69" focusable="false" viewBox="0 0 24 24" color={[undefined]} aria-hidden="true" role="presentation">
                                <path d="M15.41 16.59L10.83 12l4.58-4.59L14 6l-6 6 6 6 1.41-1.41z" />
                                <path fill="none" d="M0 0h24v24H0V0z" />
                              </svg>
                            </SvgIcon>
                          </WithStyles(SvgIcon)>
                        </KeyboardArrowLeftIcon>
                      </pure(KeyboardArrowLeftIcon)>
                    </span>
                    <h1 className="RSFBackNav-label-2">
                      Rugs
                    </h1>
                  </span>
                </Typography>
              </WithStyles(Typography)>
            </div>
          </Paper>
        </WithStyles(Paper)>
      </BackNav>
    </inject-BackNav-with-history>
  </WithStyles(inject-BackNav-with-history)>
</Provider>
`;
